Source/Core/Common/WorkQueueThread.h

```diff
@@ -4,169 +4,168 @@
-        m_items = std::queue<T>();
-      }
+    if (wait_for_completion)
+      WaitForCompletion();
 
-      m_shutdown = true;
-      m_worker_cond_var.notify_one();
+    if (m_run_thread.exchange(false, std::memory_order_relaxed))
+    {
+      m_event.Set();
+      m_thread.join();
     }
-
-    m_thread.join();
   }
 
-  // Blocks until all items in the queue have been processed (or cancelled)
-  void WaitForCompletion()
+  auto GetLockGuard()
   {
-    std::unique_lock lg(m_lock);
-    // don't check m_shutdown, because it gets set to request a shutdown, and we want to wait until
-    // after the shutdown completes.
-    // We also check m_cancelling, because we want to ensure the worker acknowledges our cancel.
-    if (m_idle && !m_cancelling.load())
-      return;
-
-    m_wait_cond_var.wait(lg, [&] { return m_idle && !m_cancelling; });
+    struct DummyLockGuard
+    {
+      // Silences unused variable warning.
+      ~DummyLockGuard() { void(); }
+    };
+
+    if constexpr (IsSingleProducer)
+      return DummyLockGuard{};
+    else
+      return std::lock_guard(m_mutex);
   }
 
-  // If the worker polls IsCanceling(), it can abort its work when Cancelling
-  bool IsCancelling() const { return m_cancelling.load(); }
+  bool IsRunning() { return m_thread.joinable(); }
 
-private:
-  void ThreadLoop()
+  void ThreadLoop(const std::string& thread_name, const std::function<void(T)>& function)
   {
-    Common::SetCurrentThreadName(m_thread_name.c_str());
+    Common::SetCurrentThreadName(thread_name.c_str());
 
-    while (true)
+    while (m_run_thread.load(std::memory_order_relaxed))
     {
-      std::unique_lock lg(m_lock);
-      while (m_items.empty())
+      if (m_items.Empty())
       {
-        m_idle = true;
-        m_cancelling = false;
-        m_wait_cond_var.notify_all();
-        if (m_shutdown)
-          return;
-
-        m_worker_cond_var.wait(
-            lg, [&] { return !m_items.empty() || m_shutdown || m_cancelling.load(); });
+        m_event.Wait();
+        continue;
       }
-      T item{std::move(m_items.front())};
-      m_items.pop();
-      lg.unlock();
 
-      m_function(std::move(item));
+      if (m_skip_work.load(std::memory_order_relaxed))
+      {
+        m_items.Clear();
+        continue;
+      }
+
+      function(std::move(m_items.Front()));
+      m_items.Pop();
     }
   }
 
-  std::function<void(T)> m_function;
-  std::string m_thread_name;
   std::thread m_thread;
-  std::mutex m_lock;
-  std::queue<T> m_items;
-  std::condition_variable m_wait_cond_var;
-  std::condition_variable m_worker_cond_var;
-  std::atomic<bool> m_cancelling = false;
-  bool m_idle = true;
-  bool m_shutdown = false;
+  Common::WaitableSPSCQueue<T> m_items;
+  Common::Event m_event;
+  std::atomic_bool m_skip_work = false;
+  std::atomic_bool m_run_thread = false;
+
+  using ProducerMutex = std::conditional_t<IsSingleProducer, std::nullptr_t, std::recursive_mutex>;
+  ProducerMutex m_mutex;
```

Comment:
This probably matters exactly never but `std::nullptr_t` is 8 bytes when a plain `bool` or `char` would only be 1 (which rounds down to effectively 0 due to alignment padding).

Reply:
I'll change it.